# Projekt 2 - ELECTRE III i PROMETHEE

#### Filip Marciniak 148148, Szymon Pasternak 148146

# Opis zbioru danych

1. Jaka jest domena problemu? 

W problemie rozważamy koszykarzy z ligi NBA grających w sezonie 2023/2024.

2. Jakie jest źródło danych?

Dane pochodzą z oficjalnej strony NBA - https://www.nba.com/stats/players/traditional.

3. Jaki jest punkt widzenia decydenta?

Jako decydent zależy nam na znalezieniu rankingu wszystkich rozważanych koszykarzy od najlepszego do najgorszego.

4. Jaka jest liczba rozważanych wariantów? Czy w oryginalnym zbiorze danych było ich więcej?

W przygotowanym zbiorze danych jest 20 wariantów. Wybrani zostali koszykarze, który uzyskują najwięcej punktów na mecz oraz zagrali przynajmniej 30 meczów. W oryginalnym zbiorze dostępnych jest 549 koszykarzy.

In [1]:
import pandas as pd
import numpy as np
pd.set_option('display.float_format', lambda x: '%.5f' % x)

Poniżej przedstawiony został pełen zbiór danych.

In [16]:
df = pd.read_csv('./Basketball_players.csv', sep=';', nrows=20)
df

PLAYER      PTS      FG%      3P%      FT%      REB  \
0               Joel Embiid 35.30000 53.30000 36.60000 88.30000 11.30000   
1               Luka Doncic 33.90000 48.70000 37.40000 78.70000  9.10000   
2     Giannis Antetokounmpo 30.70000 61.60000 28.80000 65.70000 11.50000   
3   Shai Gilgeous-Alexander 30.40000 54.00000 36.70000 87.10000  5.60000   
4              Kevin Durant 27.60000 52.90000 42.20000 85.50000  6.80000   
5          Donovan Mitchell 27.40000 46.80000 37.30000 86.20000  5.30000   
6             Jalen Brunson 27.40000 47.70000 40.10000 84.50000  3.60000   
7              Jayson Tatum 27.30000 47.20000 37.80000 83.20000  8.20000   
8              Devin Booker 27.20000 49.20000 36.40000 88.10000  4.80000   
9             Stephen Curry 26.50000 44.70000 40.30000 92.30000  4.40000   
10             De'Aaron Fox 26.50000 46.60000 36.30000 73.00000  4.50000   
11               Trae Young 26.40000 42.60000 37.10000 85.60000  2.70000   
12          Anthony Edwards 26.20000 46.40000 37.30000 82.70000  5.50000   
13             Nikola Jokic 26.10000 58.10000 35.00000 82.80000 12.20000   
14             Tyrese Maxey 25.80000 44.90000 37.90000 86.70000  3.70000   
15             LeBron James 25.30000 53.00000 40.50000 74.80000  7.30000   
16             Kyrie Irving 25.30000 49.10000 40.70000 90.50000  5.00000   
17            Anthony Davis 24.70000 55.30000 28.90000 81.50000 12.60000   
18           Damian Lillard 24.40000 42.60000 35.80000 92.30000  4.40000   
19            Julius Randle 24.00000 47.20000 31.10000 78.10000  9.20000   

        AST     TOV      PF      +/-  
0   5.70000 3.70000 2.90000  6.90000  
1   9.80000 4.00000 2.10000  3.50000  
2   6.40000 3.40000 2.90000  4.60000  
3   6.30000 2.20000 2.50000  7.90000  
4   5.20000 3.30000 1.70000  3.80000  
5   6.10000 2.90000 2.20000  7.00000  
6   6.50000 2.40000 1.90000  6.10000  
7   4.90000 2.50000 2.00000  8.80000  
8   6.90000 2.50000 3.00000  4.60000  
9   5.00000 2.80000 1.60000  2.00000  
10  5.60000 2.70000 2.60000  2.60000  
11 10.80000 4.30000 2.10000 -2.20000  
12  5.20000 3.10000 1.80000  6.50000  
13  9.00000 2.90000 2.50000  8.20000  
14  6.10000 1.70000 2.20000  3.50000  
15  8.10000 3.30000 1.20000  2.70000  
16  5.30000 1.70000 1.90000  3.70000  
17  3.60000 2.00000 2.40000  1.10000  
18  7.10000 2.50000 1.80000  5.80000  
19  5.00000 3.50000 2.70000  4.70000

5. Opisz jeden z rozważanych wariantów - jak jest oceniony na poszczególnych kryteriach, jak dobry jest na tle innych.

**Nikola Jokic**:
- PTS - 26.1 - znajduje się w drugim kwartale, nie wypada najlepiej, ale nie jest też najgorzej,
- FG% - 58.1 - na tym kryterium jest jednym z najlepszych wariantów,
- 3P% - 35.0 - na tym kryterium znajduje się w pierwszym kwartale,
- FT% - 82.2 - znajduje się w drugim kwartale,
- REB - 12.2 - wyróżnia się spośród innych wariantów dwucyfrowym wynikiem na tym kryterium, są jedynie 4 oceny dwucyfrowe,
- AST - 9.0 - na tym kryterium również jest jednym z lepszych wariantów,
- TOV - 2.9 - na tym kryterium, radzi sobie gorzej niż inne warianty, ale o niewielką wartość
- PF - 2.5 - na tym kryterium radzi sobie średnio,
- +/- - 8.2 - ostatnie kryterium zdecydowanie wyróżnia go spośród innych wariantów, uzyskuje na nim jedną z najlepszych ocen.

6. Ile rozważasz kryteriów? Czy było ich więcej w oryginalnym zbiorze danych?
   
W problemie rozważanych jest 9 kryteriów. W oryginalnym zbiorze danych było ich więcej natomiast wybrane zostały takie, które uśredniają niektóre z dostępnych oryginalnie, np. 3P% lub są najbardziej istotne, takie jak np. liczba punktów na mecz.

8. Jakie są typy poszczególnych kryteriów? Scharakteryzuj je:
Na wszystkich z kryteriów uzyskane wartości są ciągłe, ponieważ są to statystyki średnie na mecz.

- PTS - ZYSK - średnia liczba punktów na mecz
- FG% - ZYSK - procent trafionych rzutów
- 3P% - ZYSK - procent trafionych rzutów 3 punktowych
- FT% - ZYSK - procent trafionych rzutów osobistych
- REB - ZYSK - średnia liczba zbiórek na mecz
- AST - ZYSK - średnia liczba asyst na mecz
- TOV - KOSZT - średnia liczba strat na mecz
- PF - KOSZT - średnia liczba fauli na mecz
- +/- - ZYSK - średnia różnica pomiędzy punktami zdobytymi i straconymi przez drużynę w czasie kiedy gracz był obecny na boisku na mecz

Osiągane wartości:

In [17]:
df.describe()

PTS      FG%      3P%      FT%      REB      AST      TOV       PF  \
count 20.00000 20.00000 20.00000 20.00000 20.00000 20.00000 20.00000 20.00000   
mean  27.42000 49.59500 36.71000 83.38000  6.88500  6.43000  2.87000  2.20000   
std    2.98780  5.08749  3.61137  6.68026  3.11487  1.78771  0.71532  0.48232   
min   24.00000 42.60000 28.80000 65.70000  2.70000  3.60000  1.70000  1.20000   
25%   25.67500 46.55000 36.17500 80.80000  4.47500  5.20000  2.47500  1.87500   
50%   26.50000 48.20000 37.20000 85.00000  5.55000  6.10000  2.85000  2.15000   
75%   27.45000 53.07500 38.45000 87.35000  9.12500  6.95000  3.32500  2.52500   
max   35.30000 61.60000 42.20000 92.30000 12.60000 10.80000  4.30000  3.00000   

           +/-  
count 20.00000  
mean   4.59000  
std    2.68110  
min   -2.20000  
25%    3.30000  
50%    4.60000  
75%    6.60000  
max    8.80000

8. Jakie są wagi poszczególnych kryteriów?

Aby ustalić wagi danych kryteriów zastosowana została procedura SRF. Poniżej zaprezentowane zostało ustawienie kart:

<div>
    <img src="./criteria_cards.png" alt="cards" width="700"/>
</div>

`Ratio Z` ustawione zostało na 6. Po wykonaniu obliczeń wagi przedstawiają się nastepująco:

| CRITERION | WEIGHT |
| --------- | ------ |
| PTS | 6 |
| AST, REB | 3.86 |
| FG%, 3P% | 3.14 |
| TOV | 2.43 |
| FT%, +/- | 1.71 |
| PF | 1 |

In [18]:
CRITERIA = {
    'PTS': {'type': 'gain', 'k': 6, 'q': 4, 'p': 8, 'v': 12},
    'FG%': {'type': 'gain', 'k': 3.14, 'q': 3, 'p': 4, 'v': 8},
    '3P%': {'type': 'gain', 'k': 3.14, 'q': 4, 'p': 5, 'v': 11},
    'FT%': {'type': 'gain', 'k': 1.71, 'q': 3, 'p': 7, 'v': 15},
    'REB': {'type': 'gain', 'k': 3.86, 'q': 2, 'p': 5, 'v': 9},
    'AST': {'type': 'gain', 'k': 3.86, 'q': 1.5, 'p': 3, 'v': 8},
    'TOV': {'type': 'cost', 'k': 2.43, 'q': 1, 'p': 1.5, 'v': 2.5},
    'PF':  {'type': 'cost', 'k': 1, 'q': 1, 'p': 1.5, 'v': 2},
    '+/-': {'type': 'gain', 'k': 1.71, 'q': 3, 'p': 4, 'v': 8},
}

9. Czy istnieją jakieś dominacje w zbiorze wariantów? Jeśli tak, podaj je.

W zbiorze istnieją warianty zdominowane przez inne.

In [19]:
for alt_1_id in df.index:
    for alt_2_id in df.index:
        if alt_1_id == alt_2_id:
            continue

        is_at_least_as_good_on_all = all([df.loc[alt_1_id, cr] <= df.loc[alt_2_id, cr] for cr, config in CRITERIA.items() if config['type'] == 'gain']) \
               and all([df.loc[alt_1_id, cr] >= df.loc[alt_2_id, cr] for cr, config in CRITERIA.items() if config['type'] == 'cost'])
        is_better_on_one = any([df.loc[alt_1_id, cr] < df.loc[alt_2_id, cr] for cr, config in CRITERIA.items() if config['type'] == 'gain']) \
                or any([df.loc[alt_1_id, cr] > df.loc[alt_2_id, cr] for cr, config in CRITERIA.items() if config['type'] == 'cost'])
        
        if is_at_least_as_good_on_all and is_better_on_one:
            print(f"{df.loc[alt_2_id, 'PLAYER'].ljust(23)} dominates {df.loc[alt_1_id, 'PLAYER']}")

Shai Gilgeous-Alexander dominates De'Aaron Fox
Nikola Jokic            dominates Julius Randle


10. Jak teoretycznie powinien prezentować się najlepszy wariant w Twoim zbiorze danych?

Najlepszy wariant powinien osiągać bardzo dobre wartości, przede wszystkim, na kryteriach, które mają największą wagę. Wydaje się, że najbardziej wszechstronnnym graczem na kryteriach PTS, AST, REB, FG%, 3P% jest Luka Doncic lub Nikola Jokic. Każdy z nich jest jednak poniżej liderów na jednym z ważnych kryteriów. Dla Luki Doncica jest to kryterium FG%, a dla Nikoli Jokica PTS. Najlepszy gracz powinien mieć wyrównane, dobre statystyki na większości kryteriów, ponieważ wagi nie różnią się znacząco. Gdyby jednak jego statystyki na jednym z kryteriów odstawały na jego niekorzyść od reszty, to lepiej żeby to było kryterium o mniejszej wadze. W przykładzie powyżej, lepiej prezentuje się Nikola Jokic, bo pomimo niskiej wartości na kryterium PTS, które posiada najwyższą wagę, jego pozostałe statystyki (w szczególności AST i REB) prezentują się bardzo dobrze.

11. Spośród rozważanych wariantów, który wydaje się być najlepszy/lepszy od innych i dlaczego?

Najlepsi będą koszykarze, którzy mają najwięcej zdobytych punktów na mecz, ponieważ to kryterium posiada największą wagę. Najlepsi więc wydają się być Joel Embiid i Nikola Jokic.

12. Spośród rozważanych wariantów, który wydaje się być najsłabszy/ znacząco słabszy od innych i dlaczego?

Najgorsi będą koszykarze, którzy mają najmniej zdobytych punktów na mecz, ponieważ to kryterium posiada największą wagę. Najgorszy więc wydaje się być Julius Randle.